odl/tomo/operators/ray_trafo.py

```diff
@@ -273,164 +161,119 @@ def __init__(self, reco_space, geometry, variant, **kwargs):
         else:
             # proj_space was given, checking some stuff
             if not isinstance(proj_space, DiscreteLp):
-                raise TypeError('`{}` must be a `DiscreteLp` instance, '
-                                'got {!r}'.format(proj_name, proj_space))
+                raise TypeError(
+                    '`proj_space` must be a `DiscreteLp` instance, '
+                    'got {!r}'.format(proj_space))
             if proj_space.shape != geometry.partition.shape:
-                raise ValueError('`{}.shape` not equal to `geometry.shape`: '
-                                 '{} != {}'.format(proj_name, proj_space.shape,
-                                                   geometry.partition.shape))
+                raise ValueError(
+                    '`proj_space.shape` not equal to `geometry.shape`: '
+                    '{} != {}'.format(proj_space.shape,
+                                      geometry.partition.shape))
             if proj_space.dtype != reco_space.dtype:
-                raise ValueError('`{}.dtype` not equal to `{}.dtype`: '
-                                 '{} != {}'.format(proj_name, reco_name,
-                                                   proj_space.dtype,
-                                                   reco_space.dtype))
+                raise ValueError('`proj_space.dtype` not equal to '
+                                 '`reco_space.dtype`: {} != {}'
+                                 .format(proj_space.dtype, reco_space.dtype))
+
+        if reco_space.ndim != geometry.ndim:
+            raise ValueError('`reco_space.ndim` not equal to `geometry.ndim`: '
+                             '{} != {}'.format(reco_space.ndim,
+                                               geometry.ndim))
+
+        # Cache for input/output arrays of transforms
+        self.use_cache = kwargs.pop('use_cache', True)
+
+        # Check `impl`
+        impl = kwargs.pop('impl', None)
+        impl_type, self.__cached_impl = self._check_impl(impl)
+        self._impl_type = impl_type
+        self.__impl = impl.lower() if isinstance(impl, str) else impl_type
```

Comment:
This I find a bit worrying. We try very hard to make every stored attribute a fixed type, or `None` if it makes sense, in particular for those that back public properties. Never one of several possible types. That's a recipe for downstream bugs.
I'd suggest you keep `impl` to always be a `str` and look up the corresponding type in a dictionary when needed. I don't think there's a need to store the type in the class. The fewer attributes, the better.

Reply:
I changed `RayTransform.impl` so that it always returns a `str`. The reason why we currently have this hassle of storing the type is that `RayTransform` has "knowledge" of how to instantiate the implementation. Normally I wouldn't like that, but it makes sense because `RayTransform` needs to reinitiate the backend when `use_cache=False`. I guess I can also change the class so that it will directly use `type(self.__cached_impl)`, and then we'd have to make sure that there is always a `self.__cached_impl` available.

Reply:
Hm, yes, that makes sense. I'll have a closer look.

I recognize this problem from classes like `DiscreteLp` that are composed of other classes. The moment you want to "convert" the containing class, e.g., change the `dtype`, it suddenly needs to know how to instantiate its components. An annoying issue, and I haven't found a good solution yet.

Reply:
Thought about this. To get "creation knowledge" outside of the class we'd need either DI (not very ODL-style) or push it creation off to some external factory class. A bit of an overkill. Since `RayTransform` really composes and not "aggregates" `Impl`s, I don't think it is too weird for `RayTransform` to instantiate the `Impl`s. However, I don't think I'm a big fan of the root of this problem, the `use_cache` methodology. Caching/recreating the `Impl` is a way to prevent excessive memory usage after retrieving the volume data from GPU, while guaranteeing a `DiscreteLp` element. But do we not need to give users more explicit control over what is on the GPU in some future anyway? Like neural networks, that often implement `.numpy()` methods. And, if CuPy makes it into ODL, some methods may return CuPy arrays which probably shouldn't be cast into Numpy arrays by default.

At the moment I find only one example using `use_cache`. I'd think that we could either leave cache out, and people that need this can still `del ray_trafo` or reset `ray_trafo.impl`. Or maybe we'll leave it in until we find a good solution to the problems above?

Reply:
These are all good thoughts and important to keep in mind. For the moment, though, I think the current solution is (more than) acceptable. One of the points of this PR is to make the `RayTransform` extensible by using a common interface for the `Impl` classes, which makes it easy to instantiate them correctly in the caching logic.
Regarding future-proofness of the caching: ASTRA (as of now) works with host memory via copying or device memory via linking, and with the current mechanism we have the first case covered. At some point we'll modify the caching logic to account for existing GPU memory. That will require a bit of restructuring, but I don't see more cases than that.

Reply:
I agree that this is not for this PR :)